src/category_theory/sums.lean

```diff
@@ -0,0 +1,168 @@
+-/
+
+namespace category_theory
+
+universes v₁ u₁ -- declare the `v`'s first; see `category_theory.category` for an explanation
+
+open sum
+
+section
+variables (C : Type u₁) [𝒞 : category.{v₁+1} C] (D : Type u₁) [𝒟 : category.{v₁+1} D]
+include 𝒞 𝒟
+
+/--
+`sum C D` gives the direct sum of two categories.
+-/
+instance sum : category.{v₁+1} (C ⊕ D) :=
+{ hom :=
+    λ X Y, match X, Y with
+    | inl X, inl Y := X ⟶ Y
+    | inl X, inr Y := pempty
+    | inr X, inl Y := pempty
+    | inr X, inr Y := X ⟶ Y
+    end,
+  id :=
+    λ X, match X with
+    | inl X := 𝟙 X
+    | inr X := 𝟙 X
+    end,
+  comp :=
+    λ X Y Z f g, match X, Y, Z, f, g with
+    | inl X, inl Y, inl Z, f, g := f ≫ g
+    | inr X, inr Y, inr Z, f, g := f ≫ g
+    end }
+
+@[simp] lemma sum_comp_inl {P Q R : C} (f : (inl P : C ⊕ D) ⟶ inl Q) (g : inl Q ⟶ inl R) :
+  f ≫ g = (f : P ⟶ Q) ≫ (g : Q ⟶ R) := rfl
+@[simp] lemma sum_comp_inr {P Q R : D} (f : (inr P : C ⊕ D) ⟶ inr Q) (g : inr Q ⟶ inr R) :
+  f ≫ g = (f : P ⟶ Q) ≫ (g : Q ⟶ R) := rfl
+end
+
+namespace sum
+
+variables (C : Type u₁) [𝒞 : category.{v₁+1} C] (D : Type u₁) [𝒟 : category.{v₁+1} D]
+include 𝒞 𝒟
+
+/-- `inl_` is the functor `X ↦ inl X`. -/
+-- Unfortunate naming here, suggestions welcome.
+def inl_ : C ⥤ C ⊕ D :=
+{ obj := λ X, inl X,
+  map := λ X Y f, f }
+
+@[simp] lemma inl_obj (X : C) : (inl_ C D).obj X = inl X := rfl
+@[simp] lemma inl_map {X Y : C} {f : X ⟶ Y} : (inl_ C D).map f = f := rfl
+
+/-- `inr_` is the functor `X ↦ inr X`. -/
+def inr_ : D ⥤ C ⊕ D :=
+{ obj := λ X, inr X,
+  map := λ X Y f, f }
+
+@[simp] lemma inr_obj (X : D) : (inr_ C D).obj X = inr X := rfl
+@[simp] lemma inr_map {X Y : D} {f : X ⟶ Y} : (inr_ C D).map f = f := rfl
+
+/-- The functor exchanging two direct summand categories. -/
+def swap : C ⊕ D ⥤ D ⊕ C :=
+{ obj :=
+    λ X, match X with
+    | inl X := inr X
+    | inr X := inl X
+    end,
+  map :=
+    λ X Y f, match X, Y, f with
+    | inl X, inl Y, f := f
+    | inr X, inr Y, f := f
+    end }
+
+@[simp] lemma swap_obj_inl (X : C) : (swap C D).obj (inl X) = inr X := rfl
+@[simp] lemma swap_obj_inr (X : D) : (swap C D).obj (inr X) = inl X := rfl
+@[simp] lemma swap_map_inl {X Y : C} {f : inl X ⟶ inl Y} : (swap C D).map f = f := rfl
+@[simp] lemma swap_map_inr {X Y : D} {f : inr X ⟶ inr Y} : (swap C D).map f = f := rfl
+
+namespace swap
+
+/-- `swap` gives an equivalence between `C ⊕ D` and `D ⊕ C`. -/
+def equivalence : C ⊕ D ≌ D ⊕ C :=
+equivalence.mk (swap C D) (swap D C)
+  (nat_iso.of_components (λ X, eq_to_iso (by { cases X; refl })) (by tidy))
+  (nat_iso.of_components (λ X, eq_to_iso (by { cases X; refl })) (by tidy))
+
+instance is_equivalence : is_equivalence (swap C D) :=
+(by apply_instance : is_equivalence (equivalence C D).functor)
+
+/-- The double swap on `C ⊕ D` is naturally isomorphic to the identity functor. -/
+def symmetry : swap C D ⋙ swap D C ≅ functor.id (C ⊕ D) :=
+(equivalence C D).unit_iso.symm
+
+end swap
```

Comment:
Do you also want an associator (and unitor?) here?

Reply:
I've rearranged the files a little, to match the layout for products, and added the associator, and that it is an equivalence.